### ISSUE TITLE
NullReferenceException when publishing

### ISSUE THREAD

**Issue by ronaldbarendse** [umbraco/Umbraco.Cloud.Issues#126]
When publishing content with unpublished children, the following exception is thrown/shown:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Umbraco.Deploy.Cloud.LiveEditing.EventSubscriber.SendPublishedPageToClients(IContent content, IHubContext hub, UmbracoHelper helper)
   at Umbraco.Deploy.Cloud.LiveEditing.EventSubscriber.ContentService_Published(IContentService sender, PublishEventArgs`1 e)
   at Umbraco.Core.Events.TypedEventHandler`2.Invoke(TSender sender, TEventArgs e)
   at Umbraco.Core.Events.EventDefinition`2.RaiseEvent() in d:\a\1\s\src\Umbraco.Core\Events\EventDefinition.cs:line 69
   at Umbraco.Core.Events.QueuingEventDispatcher.ScopeExitCompleted() in d:\a\1\s\src\Umbraco.Core\Events\QueuingEventDispatcher.cs:line 23
   at Umbraco.Core.Events.QueuingEventDispatcherBase.ScopeExit(Boolean completed) in d:\a\1\s\src\Umbraco.Core\Events\QueuingEventDispatcherBase.cs:line 337
   at Umbraco.Core.Scoping.Scope.<>c__DisplayClass72_0.<RobustExit>b__1() in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 436
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions) in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 474
   at Umbraco.Core.Scoping.Scope.TryFinally(Int32 index, Action[] actions) in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 478
   at Umbraco.Core.Scoping.Scope.RobustExit(Boolean completed, Boolean onException) in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 422
   at Umbraco.Core.Scoping.Scope.DisposeLastScope() in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 402
   at Umbraco.Core.Scoping.Scope.Dispose() in d:\a\1\s\src\Umbraco.Core\Scoping\Scope.cs:line 365
   at Umbraco.Core.Services.Implement.ContentService.SaveAndPublish(IContent content, String culture, Int32 userId, Boolean raiseEvents) in d:\a\1\s\src\Umbraco.Core\Services\Implement\ContentService.cs:line 907
```

After opening the children and publishing them first (bottom-up), no exception is thrown. So it's probably expecting all children to be published, which might not be the case: the content doesn't exist in the cache and `null` is returned instead (which is not correctly checked)...

Reproduction
------------

### Bug summary
Exception thrown after publishing content with unpublished children.

### Specifics
Umbraco 8.1.5
Umbraco Deploy 3.1.3

Tested on a local workspace using single and batch publishing of content (using list view).

### Steps to reproduce
- Create a tree of content by only saving
- Publish the highest content in the tree (the root)

### Expected result
Publish all content and not throw a `NullReferenceException`.

### Actual result
An exception thrown and only the first content item gets published.

---
_This item has been added to our backlog [AB#4423](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/4423)_

**Comment by ronaldbarendse** [comment]
After figuring out what `SendPublishedPageToClients` does, it looks like this is the culprit: `helper.Content(content.Id).Url()` (like expected, fetching an unpublished item from the cache). Getting the URL can also be done without using the `UmbracoHelper` by directly calling `UmbracoContext.UrlProvider.GetUrl()`.

**Comment by hfloyd** [comment]
I am seeing the same error message, but it happens when I create a new content node and do the initial "Save & Publish"

Umbraco version 8.3.0

**Comment by ronaldbarendse** [comment]
We're 3 months in and it doesn't look like this bug is fixed yet, even though a new minor version of Umbraco Deploy 3.3.0 is just released with new features 🤷‍♂

**Comment by thomashdk** [comment]
I getting the same errors now on Umbraco 8.4 and Ubmraco.Deploy 3.3.0..

**Comment by warrenbuckley** [comment]
@thomashdk & @ronaldbarendse I have tried with a brand new Umbraco Cloud V8 project and I am unable to reproduce directly on Cloud or when the site is cloned down locally.

Are you able to give me any more information when you are doing this on a site with Umbraco 8.4.0 & Deploy 3.3.0 please?

## Versions:
Umbraco: 8.4.0
Deploy: 3.3.0

## My Repro Steps
* Create one doctype
* Create three content nodes and only save them
* Publish with descendants from root node (a) and ensure checkbox enabled for unpublished nodes

## Content Structure

* a
   * b
        * c

**Comment by warrenbuckley** [comment]
## Update
OK a colleague @HalldorLyngmo has given me some insight into this and told me that I need to `Save & Publish` the root node, as opposed to `Publish with Descendants`

**Comment by nul800sebastiaan** [comment]
Apologies for not updating this issue earlier, I just saw that we released Deploy 3.3.1 in late January and it addresses this issue, so I'll close this one as fixed!
